pandas/tools/merge.py

```diff
@@ -740,21 +763,35 @@ def _get_merge_keys(self):
                         right_keys.append(rk)
                         join_names.append(None)  # what to do?
                     else:
-                        right_keys.append(right[rk]._values)
-                        join_names.append(rk)
+                        if rk is not None:
+                            right_keys.append(right[rk]._values)
+                            join_names.append(rk)
+                        else:
+                            # kludge for merge_asof(right_index=True)
+                            right_keys.append(right.index.values)
+                            join_names.append(right.index.name)
                 else:
                     if not is_rkey(rk):
-                        right_keys.append(right[rk]._values)
-                        if lk == rk:
+                        if rk is not None:
+                            right_keys.append(right[rk]._values)
+                        else:
+                            # kludge for merge_asof(right_index=True)
+                            right_keys.append(right.index.values)
+                        if lk is not None and lk == rk:
                             # avoid key upcast in corner case (length-0)
                             if len(left) > 0:
                                 right_drop.append(rk)
                             else:
                                 left_drop.append(lk)
                     else:
                         right_keys.append(rk)
-                    left_keys.append(left[lk]._values)
-                    join_names.append(lk)
+                    if lk is not None:
+                        left_keys.append(left[lk]._values)
+                        join_names.append(lk)
+                    else:
+                        # kludge for merge_asof(left_index=True)
```

Comment:
This is unique to `pd.merge_asof()` because `on` and `by` are separate parameters. So a user could, for example, request `left_index` and `left_by`.

In a regular `pd.merge()`, users cannot specify both `left_index` and `left_on`. (Instead, users have a `MultiIndex`). That means the `left_on` in `_get_merge_keys()` is always empty in a `pd.merge()`, but a `pd.merge_asof(left_index=True, left_by=...)` will result in a `left_on` array with a `None` in the middle of it. See `_validate_specification()` for where this happens.

I put the kludge in there to handle this case. So perhaps it isn't a kludge at all and I should just put the above explanation at the top of the function. What do you think?